content/guides/docker-concepts/building-images/multi-stage-builds.md

```diff
@@ -0,0 +1,327 @@
+
+     ```dockerfile
+     COPY src ./src
+     ```
+
+
+ 7. Set the default command to be executed when the container starts. This command instructs the container to run the Maven wrapper (`./mvnw`) with the `spring-boot:run` goal, which will build and execute your Spring Boot application.
+
+     ```dockerfile
+     CMD ["./mvnw", "spring-boot:run"]
+     ```
+
+    And with that, you should have the following Dockerfile:
+
+    ```dockerfile 
+    FROM eclipse-temurin:21.0.2_13-jdk-jammy
+    WORKDIR /app
+    COPY .mvn/ .mvn
+    COPY mvnw pom.xml ./
+    RUN ./mvnw dependency:go-offline
+    COPY src ./src
+    CMD ["./mvnw", "spring-boot:run"]
+    ```
+
+### Build the container image
+
+
+ 1. Execute the following command to build the Docker image:
+
+
+    ```console
+    $ docker build -t spring-helloworld .
+    ```
+
+ 2. Check the size of the Docker image by using the `docker images` command:
+
+    ```console
+    $ docker images
+    ```
+
+    Doing so will produce output like the following:
+
+    ```console
+    REPOSITORY          TAG       IMAGE ID       CREATED          SIZE
+    spring-helloworld   latest    ff708d5ee194   3 minutes ago    880MB
+    ```
+
+
+    Based on this output, you can see that your image is 880MB in size! In this image, you have the full JDK, Maven toolchain, and more. In production, you don’t need that in your final image.
+
+
+### Run the Spring Boot application
+
+1. Now that you have an image built, it's time to run the container.
+
+    ```console
+    $ docker run -d -p 8080:8080 spring-helloworld
+    ```
+
+    You'll then see output similar to the following in the container log:
+
+    ```plaintext
+    [INFO] --- spring-boot:3.3.0-M3:run (default-cli) @ spring-boot-docker ---
+    [INFO] Attaching agents: []
+     .   ____          _            __ _ _
+     /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
+     ( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
+     \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
+     '  |____| .__|_| |_|_| |_\__, | / / / /
+      =========|_|==============|___/=/_/_/_/
+
+     :: Spring Boot ::             (v3.3.0-M3)
+
+     2024-04-04T15:36:47.202Z  INFO 42 --- [spring-boot-docker] [           main]       
+     c.e.s.SpringBootDockerApplication        : Starting SpringBootDockerApplication using Java    
+     21.0.2 with PID 42 (/app/target/classes started by root in /app)
+     ….
+     ```
+
+
+2. Access your “Hello World” page through your web browser at [http://localhost:8080](http://localhost:8080), or via this curl command:
+
+    ```console
+    $ curl localhost:8080
+    Hello World
+    ```
+
+### Use multi-stage builds
+
+1. Consider the following Dockerfile:
+
+    ```dockerfile
+    FROM eclipse-temurin:21.0.2_13-jdk-jammy as builder
+    WORKDIR /opt/app
+    COPY .mvn/ .mvn
+    COPY mvnw pom.xml ./
+    RUN ./mvnw dependency:go-offline
+    COPY ./src ./src
+    RUN ./mvnw clean install
+    FROM eclipse-temurin:21.0.2_13-jre-jammy as final
```

Comment:
The documentation for this image recommends using jlink to create a runtime. Maybe we should include that in this example, with a `FROM debian:bookworm-slim`

Reply:
@dvdksn Did you test it with debian:bookworm-slim?  Also, can you point me to the documentation link?

Reply:
I tried to replace `eclipse-temurin:21.0.2_13-jre-jammy` with `debian:bookworm-slim` but it didn't work. Also, there is an open issue around it https://github.com/adoptium/installer/issues/766

I would rather stick to eclipse-temurin:21.0.2_13-jre-jammy. What do you think?

Reply:
Hi, I didn't try it myself, but it's called out in the eclipse-temurin DOI readme https://hub.docker.com/_/eclipse-temurin

The example there uses `buster` though, maybe there's some issue with bookworm specifically, I don't know. But it's probably fine to keep the `jre` tag of the eclipse-temurin image as an example as well. Just wanted to call it out as it goes against the recommendations of the DOI

Reply:
Thanks for the info. How about capturing this as a note "JRE images are available for all versions of Eclipse Temurin but it is recommended that you produce a custom JRE-like runtime using jlink for your production use. [Refer to this page](https://hub.docker.com/_/eclipse-temurin) for more information"

Reply:
I have added a note. Can you please check it once?

Reply:
Looks good to me